File: 08-Middle-of-Linked-List/middleNode.go
package main

type Node struct {
	data int
	next *Node
}

func middleNode(head *Node) *Node {
	slow := head
	fast := head
	for fast != nil && fast.next != nil {
		fast = fast.next.next
		slow = slow.next
	}

	return slow
}

func main() {
	input := []int{1, 2, 3, 4, 5}
	head := &Node{}
	cur := head
	for i := range input {
		cur.data = input[i]
		if len(input)-i != 1 {
			cur.next = &Node{}
			cur = cur.next
		}
	}
	res := middleNode(head)
	println(res.data)
	// cur1 := head
	// for cur1 != nil {
	// 	println(cur1.data)
	// 	cur1 = cur1.next
	// }
}

File: 08-Middle-of-Linked-List/middleNode.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def minddleNode(head: Node) -> Node:
    slow = head
    fast = head
    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next
    return slow


def minddleNode1(head: Node) -> Node:
    length = 0
    cur = head
    while cur:
        length += 1
        cur = cur.next

    result = head
    for i in range(length // 2):
        result = result.next
    return result


if __name__ == "__main__":
    all_nodes = [1, 2, 3, 4, 5]
    head = Node(None)

    cur = head
    for i in all_nodes:
        cur.data = i
        if i <= len(all_nodes) - 1:
            cur.next = Node(None)
            cur = cur.next

    res = minddleNode(head)
    print("minddleNode:", res.data)

    print(minddleNode1(head).data)
    # cur1 = head
    # while cur1:
    #     print(cur1.data)
    #     cur1 = cur1.next

    # all_nodes = [1, 2, 3, 4, 5, 6]
    # head = Node(None)

    all_nodes = [1, 2, 3, 4, 5, 6]
    head = Node(None)

    cur = head
    for i in all_nodes:
        cur.data = i
        if i <= len(all_nodes) - 1:
            cur.next = Node(None)
            cur = cur.next

    # cur1 = head
    # while cur1:
    #     print(cur1.data)
    #     cur1 = cur1.next
    res = minddleNode(head)
    print("minddleNode:", res.data)
